### PR TITLE
Executor

### LINKED ISSUE #25: Rework execute binary opcode functions to avoid copy-paste
Maybe implement it in a more general way like 
```c++
static void executeBinaryOp(const Instruction &inst, State &state, OpType op)
```
due to copy-paste code

_Originally posted by @ZVasilii in https://github.com/106-inc/sim2022/pull/11#discussion_r1008704870_

### LINKED ISSUE #26: Implement exeption handling of division by zero case
Division by zero case must be handled in some way

_Originally posted by @ZVasilii in https://github.com/106-inc/sim2022/pull/11#discussion_r1008730283_

### LINKED ISSUE #28: Support I-type and cond branches
Previous: https://github.com/106-inc/sim2022/issues/7

On this iteration i suggest implement such instructions:

- [ ] I-type for RV32I
- [ ] conditional branches for RV32I (it wasn't implemented on previous stage)

### PR DESCRIPTION
**Pull request by Bryanskiy**
related to https://github.com/106-inc/sim2022/issues/28

### CODE REVIEW

**Comment by Bryanskiy** [comment]
#25 and #26 also fixed.

File: src/executor/executor.cc [hunk 1]
```diff
@@ -2,79 +2,205 @@
...
-  state.npc =
-      setBit<0, 0>(applyOffset(rs1, inst.imm)); // setting the least-significant
-                                                // bit of the result to zero.
+template <std::integral T> static T executeSR(T lhs, T rhs) {
+  return lhs >> rhs;
 }
 
-static void executeECALL(const Instruction &, State &state) {
-  state.complete = true;
+template <typename Func>
```
Comment (derzhavin3016) [inline src/executor/executor.cc]: I think it's better to specify concept here

File: src/executor/executor.cc [hunk 2]
```diff
@@ -2,79 +2,205 @@
...
+  auto rs1 = state.regs.get(inst.rs1);
+  auto rs2 = state.regs.get(inst.rs2);
+  state.regs.set(inst.rd, op(rs1, rs2));
 }
 
-static void executeLW(const Instruction &inst, State &state) {
-  RegVal rs1 = state.regs.get(inst.rs1);
-  RegVal word = state.mem.loadWord(applyOffset(rs1, inst.imm));
-  state.regs.set(inst.rd, word);
+template <typename Func>
```
Comment (derzhavin3016) [inline src/executor/executor.cc]: Concept

File: src/executor/executor.cc [hunk 3]
```diff
@@ -2,79 +2,205 @@
...
-  state.regs.set(inst.rd, rs1 * rs2);
+void Executor::execute(const Instruction &inst, State &state) const {
+  execMap_.at(inst.type)(inst, state);
 }
 
-static void executeDIV(const Instruction &inst, State &state) {
-  RegVal rs1 = state.regs.get(inst.rs1);
-  RegVal rs2 = state.regs.get(inst.rs2);
-  state.regs.set(inst.rd, rs1 / rs2);
+template <typename Func>
```
Comment (derzhavin3016) [inline src/executor/executor.cc]: Concept

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

**Comment by derzhavin3016** [comment]
Closes #25 
Closes #26

**Review by derzhavin3016 — APPROVED** [review]
LGMT

**Pull request merged**